nsxt/resource_nsxt_upgrade_run.go

```diff
@@ -0,0 +1,631 @@
+	}
+
+	runPostcheck(upgradeClientSet.UpgradeClient, d)
+
+	d.SetId(id)
+	return resourceNsxtUpgradeRunRead(d, m)
+}
+
+func prepareUpgrade(upgradeClientSet *upgradeClientSet, d *schema.ResourceData) error {
+	for i := range upgradeComponentList {
+		component := upgradeComponentList[i]
+		// Customize MP upgrade is not allowed
+		if component == mpUpgradeGroup {
+			continue
+		}
+
+		status, err := getUpgradeStatus(upgradeClientSet.StatusClient, &component)
+		if err != nil {
+			return err
+		}
+		// Cannot modify succeed component upgrade.
+		if status == model.ComponentUpgradeStatus_STATUS_SUCCESS {
+			log.Printf("[WARN] %s upgrade is already succeed. Any changes on it will be ignored.", component)
+			continue
+		}
+		if status == model.ComponentUpgradeStatus_STATUS_IN_PROGRESS {
+			upgradeClientSet.PlanClient.Pause()
+		}
+		err = waitUpgradeForStatus(upgradeClientSet, &component, inFlightComponentUpgradeStatus, staticComponentUpgradeStatus)
+		if err != nil {
+			return err
+		}
+
+		// Call reset regardless, because we don't know if UpgradeUnitGroup or UpgradePlanSetting has been changed.
+		err = upgradeClientSet.PlanClient.Reset(component)
+		if err != nil {
+			return err
+		}
+
+		err = updateUpgradeUnitGroups(upgradeClientSet, d, component)
+		if err != nil {
+			return err
+		}
+
+		err = updateComponentUpgradePlanSetting(upgradeClientSet.SettingClient, d, component)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Get component upgrade status. Using nil component for overall upgrade status.
+func getUpgradeStatus(statusClient upgrade.StatusSummaryClient, component *string) (string, error) {
+	status, err := statusClient.Get(component, nil, nil)
+	if err != nil {
+		return "", err
+	}
+	if component == nil {
+		return *status.OverallUpgradeStatus, nil
+	}
+	for _, componentStatus := range status.ComponentStatus {
+		if *componentStatus.ComponentType == *component {
+			return *componentStatus.Status, nil
+		}
+	}
+	return "", fmt.Errorf("fail to retrieve upgrade status of %s component", *component)
+}
+
+// Wait component upgrade status to become target status. Using nil component for overall upgrade status.
+func waitUpgradeForStatus(upgradeClientSet *upgradeClientSet, component *string, pending, target []string) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: pending,
+		Target:  target,
+		Refresh: func() (interface{}, string, error) {
+			status, err := getUpgradeStatus(upgradeClientSet.StatusClient, component)
+			if component != nil && *component == mpUpgradeGroup && isServiceUnavailableError(err) {
+				// After MP upgrade is completed, NSXT will restart and service_unavailable error will be received here.
+				return model.ComponentUpgradeStatus_STATUS_SUCCESS, model.ComponentUpgradeStatus_STATUS_SUCCESS, nil
+			}
+			if err != nil {
+				return nil, model.ComponentUpgradeStatus_STATUS_FAILED, err
+			}
+			log.Printf("[DEBUG] Current upgrade status: %s", status)
```

Comment:
Do we get any error details in case status is Failed?

Reply:
Yes. In this case, it will be better to append the status details in the error msg.